Add FhirError component to Gluestack-ui renderer

nickvallee opened the pull request:
#180

- Alert component does not follow Gluestack-ui docs exactly because the Alert component is not meant to have buttons, so I made built it to support the retry button.
- Note: We should add a line in the future Gluestack docs that we require the use of the `  <GluestackUIProvider config={gluestackUIConfig}>` provider, because this FhirError component is using props that exist only on the styled version of the components (e.g Alert prop action is a styled prop) If we want to support unstyled gluestack components, then we'll have to build things diffferently. I'll add this to the Issue

<img width="387" alt="Screenshot 2024-01-19 at 9 47 33 AM" src="https://github.com/bonfhir/bonfhir/assets/8885263/ea099b9e-dc0c-4c65-bbb2-51eea4b356da">
<img width="381" alt="Screenshot 2024-01-19 at 9 48 08 AM" src="https://github.com/bonfhir/bonfhir/assets/8885263/ce738f87-5667-40de-bd91-971cfbaa270b">

nickvallee committed:
Add FhirError componenet to Gluestack-ui renderer

nickvallee commented:
I think it is failing because there are type errors that will only be resolved by the ` <GluestackUIProvider config={gluestackUIConfig}>` being added to the code. I suppose we might have to add it to our package somewhere

will likely have to add 

```
    "@gluestack-style/react": "^1.0.41",
    "@gluestack-ui/config": "^1.0.12",
```

julienblin commented on packages/gluestack-ui/src/r5/feedback/fhir-error.tsx:
```diff
@@ -0,0 +1,74 @@
+import { FhirErrorRendererProps } from "@bonfhir/react/r5";
+// importing from types because Alert type has identical name in themed lib
+import { IAlertProps } from "@gluestack-ui/alert/lib/typescript/types";
```
I would suggest using `ComponentProps<typeof Alert>` instead of reaching out deep into the import path to pull out `IAlertProps` - as it might be more fragile to the library updates.

nickvallee replied:
Oh nice,  I didn't know about that. Good to know!

julienblin reviewed (dismissed):
Great - thank you!
I added a suggestion that may help you with your type issues.

nickvallee commented:
@julienblin after adding your suggestion, `pnpm check` still fails because there are props on the elements that are only exist ` <GluestackUIProvider config={gluestackUIConfig}>` has been added to the app as a provider. So, it works just fine once you have imported into an expo app, but in bonfhir it throws type errors.

should we:
1. Add this provider somewhere in our code
2. add a ts-ignore line for every prop that throws a ts-error. (e.g. `action="error"` or `pt="$1"`).
3. refactor so that we are using only unstyled versions of gluestack components, and then stick with that policy

...maybe there is another option.

I would assume we want to use their styled components, but maybe I'm wrong.

julienblin commented:
> @julienblin after adding your suggestion, `pnpm check` still fails because there are props on the elements that are only exist ` <GluestackUIProvider config={gluestackUIConfig}>` has been added to the app as a provider. So, it works just fine once you have imported into an expo app, but in bonfhir it throws type errors.
> 
> should we:
> 
> 1. Add this provider somewhere in our code
> 2. add a ts-ignore line for every prop that throws a ts-error. (e.g. `action="error"` or `pt="$1"`).
> 3. refactor so that we are using only unstyled versions of gluestack components, and then stick with that policy
> 
> ...maybe there is another option.
> 
> I would assume we want to use their styled components, but maybe I'm wrong.

I think your assumption is right - we'll stick with the styled components for the moment.
It's interesting their approach to typings given what they want to achieve, I've never seen it done like that!

I tried the following:

1. Add the `@gluestack-ui/config` package as a dev dependency to the `@bonfhir/gluestack-ui` package: `pnpm add -D @gluestack-ui/config`

2. Drop the following file in the project: `src/r5/gluestack.d.ts`:

```typescript
import {
  IComponents,
  IConfig,
  componentsConfig,
  gluestackUIConfig,
} from "@gluestack-ui/config";

declare module "@gluestack-ui/themed" {
  interface UIConfig
    extends Omit<typeof gluestackUIConfig, keyof IConfig>,
      IConfig {}
  interface UIComponents
    extends Omit<typeof componentsConfig, keyof IComponents>,
      IComponents {}
}
```

This actually reproduces the global declaration that gets processed when importing the config in the provider.

This kinda "works" but it explodes the build time - it now takes about 4 minutes(!) to build the package. This is too long.
I'll see if I can narrow it down somehow.

Commit:
Fix type issues with gluestack

Typing is weird in gluestack. this adds the necessaru typings to work with the LSP and the noEmit typecheck.
However, it destroys(!) the build, so I have to exclude it when building and ignore the warnings.

julienblin commented:
@nickvallee I think I found a solution by excluding the declaration file at build time. This is not great but seems to work.
I just pushed the changes in your branch - let me know if that makes sense.
